fs.go

```diff
@@ -0,0 +1,130 @@
+package books
+
+import (
+	"io"
+	"log"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+func TruncateFilename(fn string) string {
+	var lst []string
+	dirs, fn := path.Split(fn)
+	if dirs != "" {
+		dirs = strings.TrimRight(dirs, string(os.PathSeparator))
+		lst = strings.Split(dirs, string(os.PathSeparator))
+	}
+	for i, f := range lst {
+		if len(f) <= 255 {
+			continue
+		}
+
+		l := len(f)
+		if l > 255 {
+			l = 255
+		}
+		lst[i] = f[:l]
+	}
+	if len(fn) > 250 {
+		ext := path.Ext(fn)
+		nameLen := 250 - len(ext)
+		fn = fn[:nameLen] + ext
+	}
+	lst = append(lst, fn)
+	return strings.Join(lst, string(os.PathSeparator))
+}
+
+// moveOrCopyFile moves or copies a file from book.OriginalFilename to book.CurrentFilename, relative to the configured books root.
+// All necessary directories to make the destination valid will be created.
+func moveOrCopyFile(origName, newName string, move bool) error {
+	err := os.MkdirAll(path.Dir(newName), 0755)
+	if err != nil {
+		return err
+	}
+
+	if move {
+		err = moveFile(origName, newName)
+	} else {
+		err = copyFile(origName, newName)
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// copyFile copies a file from src to dst, setting dst's modified time to that of src.
+func copyFile(src, dst string) (e error) {
+	fp, err := os.Open(src)
+	if err != nil {
+		return errors.Wrap(err, "Copy file")
+	}
+	defer fp.Close()
+
+	st, err := fp.Stat()
+	if err != nil {
+		return errors.Wrap(err, "Copy file")
+	}
+
+	fd, err := os.Create(dst)
```

Comment:
This whole method copies a file. Probably better to wrap with "create destination file."